purepyvmware/datastores.py

```diff
@@ -0,0 +1,329 @@
+            ds_name (str): Requested name of the datastore.
+            ds_size (int): Size in GB the rquested datastore should be.
+            flasharray (purestorage.FlashArray): Existing session to FlashArray.
+            vmfs_version (int): Version of VMFS datastore to be created.
+
+        Returns:
+            datastore (vim.Datastore): Newly created datastore object.
+
+        Raises:
+            ValueError:
+                - If cluster resource is not found within vCenter.
+                - If no ESXi hosts are found (empty cluster).
+                - If no connected ESXi hosts are found within the cluster.
+                - If disk device is not found on ESXi host.
+        """
+        hgroup, connected_esxi_hosts = Datastores.verify_vsphere_cluster(cluster_name, flasharray)
+
+        if connected_esxi_hosts:
+            esxi_host = connected_esxi_hosts[0]
+            fa_volume = flasharray.create_volume(ds_name, ds_size)
+            flasharray.connect_hgroup(hgroup, fa_volume['name'])
+            esxi_host.configManager.storageSystem.RescanAllHba()
+        else:
+            raise ValueError(f'No connected hosts found for cluster \'{cluster_name}\'.')
+
+        vol_serial_num = fa_volume['serial'].lower()
+        host_dssystem = esxi_host.configManager.datastoreSystem
+        avail_disks = host_dssystem.QueryAvailableDisksForVmfs()
+
+        # Confirm the selected FlashArray volume is seen by the ESXi host.
+        device_path = datastore_utils.get_device_path(avail_disks, vol_serial_num)
+
+        if not device_path:
+            disk_msg = f'Unable to find specified device on {esxi_host.name}.'
+            raise ValueError(disk_msg)
+
+        vmfs_creation_options = host_dssystem.QueryVmfsDatastoreCreateOptions(device_path)
+        vmfs_creation_options[0].spec.vmfs.volumeName = ds_name
+
+        if vmfs_version:
+            vmfs_creation_options[0].spec.vmfs.majorVersion = vmfs_version
+
+        datastore = host_dssystem.CreateVmfsDatastore(vmfs_creation_options[0].spec)
+
+        Datastores.rescan_esxi_storage(connected_esxi_hosts)
+
+        return datastore
+
+
+class VvolDatastores(Datastores):
+    """Maintenance of VMware vVol datastores associated with Pure Storage FlashArrays."""
+
+    def __init__(self, sms_instance, vsphere_content, flasharray):
+        super(VvolDatastores, self).__init__(vsphere_content, flasharray)
+        self.sms_instance = sms_instance
+
+    def create_vvol_datastore(self, cluster_name, ds_name, flasharray, pe_name=None):
+        """Create a new vVol datastore on a Pure Storage FlashArray and connect it to the requested vSphere cluster.
+
+        Args:
+            cluster_name (str): Name of the cluster to create datastore in.
+            ds_name (str): Requested name of the datastore.
+            flasharray (purestorage.FlashArray): Existing session to FlashArray.
+            pe_name (optional, str): Name of the protocol endpoint to be used on a FlashArray.
```

Comment:
some of these short names are confusing to me. The docstring helps to identify what it is, but just looking at the args "pe_name" doesn't really mean anything.

Maybe go with a longer variable name for ease of use?

e.g. `protocol_endpoint` instead of `pe_name` ?